pkg/arvo/app/btc-node-hook.hoon

```diff
@@ -0,0 +1,354 @@
+::
+/-  *btc-node-hook, *btc-node-store, sole
+/+  default-agent, sole, lib=btc-node-json, verb
+::
+=>  |%
+    +$  card  card:agent:gall
+    +$  versioned-state
+      $%  [%0 state-zero]
+      ==
+    ::
+    +$  state-zero
+      $:  endpoint=@t
+          headers=header-list:http
+      ==
+    --
+::
+=|  state-zero
+=*  state  -
+::  Main
+::
+%+  verb  |
+^-  agent:gall
+=<  |_  =bowl:gall
+    +*  this      .
+        btc-core  +>
+        bc        ~(. btc-core bowl)
+        def       ~(. (default-agent this %|) bowl)
+    ::
+    ++  on-init
+      ^-  (quip card _this)
+      :-  ~
+      %=  this
+        endpoint  ''
+        headers   ~
+      ==
+    ::
+    ++  on-save   !>(state)
+    ++  on-load
+      |=  old=vase
+      `this(state !<(state-zero old))
+    ::
+    ++  on-poke
+      |=  [=mark =vase]
+      ^-  (quip card _this)
+      |^
+      ?+    mark    (on-poke:def mark vase)
+          %btc-node-hook-action   (hook-action !<(btc-node-hook-action vase))
+          %btc-node-hook-command  (hook-command !<(btc-node-hook-command vase))
+      ==
+      ::
+      ++  hook-action
+        |=  act=btc-node-hook-action
+        ^-  (quip card _this)
+        =/  body=request:rpc:jstd
+          (request-to-rpc:btc-rpc:lib act)
+        =/  req=request:http
+          :*  %'POST'
+              (endpoint-url:bc act)
+              headers
+              =,  html
+              %-  some
+                %-  as-octt:mimes
+                  (en-json (request-to-json:rpc:jstd body))
+          ==
+        =/  out  *outbound-config:iris
+        :_  this
+        [%pass /[(scot %da now.bowl)] %arvo %i %request req out]~
+      ::
+      ++  hook-command
+        |=  comm=btc-node-hook-command
+        ^-  (quip card _this)
+        ?+    -.comm  ~|  [%unsupported-hook-command -.comm]  !!
+            %credentials
+          :_  this(endpoint url.comm, headers heads.comm)
+          [%pass / %arvo %d %flog [%text "credentials updated..."]]~
+        ==
+      --
+    ::
+    ++  on-watch  on-watch:def
+    ++  on-leave  on-leave:def
+    ++  on-peek   on-peek:def
+    ++  on-agent  on-agent:def
+    ++  on-arvo
+      |=  [=wire =sign-arvo]
+      ^-  (quip card _this)
+      =*  response  client-response.sign-arvo
+      =^  cards  state
+        ?+    +<.sign-arvo    (on-arvo:def wire sign-arvo)
+            %http-response    (http-response:bc wire response)
+        ==
+      [cards this]
+    ::
+    ++  on-fail   on-fail:def
+    --
+::
+|_  =bowl:gall
+::
+++  httr-to-rpc-response
+  |=  hit=httr:eyre
+  ^-  response:rpc:jstd
+  ~|  hit
+  ?.  ?=($2 (div p.hit 100))
+    fail+hit
+  =/  a=json  (need (de-json:html q:(need r.hit)))
+  =,  dejs-soft:format
+  ^-  response:rpc:jstd
+  =;  dere
+    =+  res=((ar dere) a)
+    ?~  res  (need (dere a))
+    [%batch u.res]
+  |=  a=json
+  ^-  (unit response:rpc:jstd)
+  =/  res=(unit [@t json])
+    ::  TODO  breaks when no id present
+    ::
+    ((ot id+so result+some ~) a)
```

Comment:
Yeah, those are stupid cases to deal with. You'll want to parse for both individually, I guess? Can always just revert to doing `?>  ?=(%o a)` and then handling the `p.a` map directly.

(`a` is a terrible name and you know it. (; )

Reply:
This is not my code you honor! :D
I just realized that I removed the comment that pointed to its [source](https://github.com/urbit/arvo/pull/973/commits/be296f6897ca6c96225d844912d7e92ea93ea75a#diff-32b4dc798ca3aa3aa5c9fbd963627d4eR1941) but it's still in my own repo:
https://github.com/yosoyubik/urbitcoin/blob/4a75a1b0736d7e02271fa7165365fcb7516b7adf/src/app/btc-node-hook.hoon#L287

Reply:
I think I'll end up using this:

```
  =/  res=[id=(unit @t) res=(unit json)]
    %.  jon
    =,  dejs:format
    =-  (ou -)
    :~  ['id' (uf ~ (mu so))]
        ['result' (un some)]
    ==
```
If id or result are not present you'll get units.